FSD and Wave coupling developments

erinethomas opened the pull request:
This PR is not intended to be merged yet - but is simply to get feedback on the on-going "icepack merge" developments for adding Floe Size Distribution code to MPAS-SI. There are many modifications needed for FSD and waves - and I would love feedback before the modifications become too large to assess easily. 

Final comment: this code is currently being tested for full coupling.

Commit:
updated .gitmodules to point ww3 and icepack to erinthomas github directory

Commit:
initial modificaiton for wave ice coupling:
added definitions of _w2x_ _x2w_ _i2x_ and _x2i_ for wave-ice coupling
modification currenlty commented out until ready for testing)

Commit:
Added ice floe size to ix2 and x2w vars
added wave spectrum to w2x and x2i vars
(note- variable commented out for now and array have not been filled with data)

Commit:
added options to allow wave coupling with MPASSi (as a command line compile option)
also updated the column_waveFracture within mpas_seaice_icepack.F (still in progress)

Commit:
Modifications to namelist setup - to allow wave-ice coupling and define the number of wave frequencies (which is set in Wave model)

Commit:
undo chnages made to add wave information to mpassi namelists.

Commit:
varibles for wave coupling added to sea ice registry (icefloe size, wave spectrum)
constants defined by WW3 needed by mpassi defined in seq_infordata_mod

Commit:
commenting fsd and wave subroutine calls out - to test seq calls

Commit:
commenting out mods to test.

Commit:
commenting out for testing

Commit:
correct compset name for DTEST- with WW3

Commit:
adding init_column_fsd subroutine to mpas_seaice_icepack.F and starting to include FSD variables to Registry.
Also changed flag name for wave-ice stuff to "config_use_column_waves"

Commit:
declare config_use_column_waves type

Commit:
several typos fixed in new fsd init subroutine.
Code compiles correctly now.

Commit:
add fsd flag to column tracers

Commit:
typo fix

Commit:
added call to icepack_init_fsd. compile correctly

Commit:
add vars for FSD to registry

Commit:
starting to add fsd tracer variable infrastructure

Commit:
bug fixes to fsd tracer code

Commit:
bugs/typo fixes in fsd tracer codes

Commit:
bug fixes/typos in tracer code

Commit:
typo fixes on tracer code

Commit:
update fsd tracer code, and init_fsd

Commit:
update fsd initalization to better match CICE code.

Commit:
final bug fixes to icepack init_fsd

Commit:
updates to icepacks initalization for fsd

Commit:
minor corrections to match variable names in registry.

eclare108213 commented on components/mpas-seaice/bld/namelist_files/namelist_defaults_mpassi.xml:
```diff
@@ -166,14 +168,14 @@
 <config_use_column_itd_thermodynamics prognostic_mode="prescribed">false</config_use_column_itd_thermodynamics>
 <config_use_column_ridging>true</config_use_column_ridging>
-<config_use_column_snow_tracers>true</config_use_column_snow_tracers>
+<config_use_column_snow_tracers>false</config_use_column_snow_tracers>
```
It's fine for you to change the default values in namelist_defaults_mpassi.xml for your own use, but these will need to be changed back for your ultimate PR.  They can be changed for individual sandboxes using the setup scripts.

eclare108213 commented on components/mpas-seaice/bld/namelist_files/namelist_definition_mpassi.xml:
```diff
@@ -920,6 +935,37 @@ Valid values: true or false
+
+<entry id="config_xfreq" type="real"
+	category="wave_config" group="wave_config">
+Frequency mulitpilcation factor used by the wave model.
```
spelling!

eclare108213 commented on driver-mct/shr/seq_infodata_mod.F90:
```diff
@@ -230,6 +230,7 @@ MODULE seq_infodata_mod
      integer(SHR_KIND_IN)    :: iac_nx          ! nx, ny of "2d" grid
      integer(SHR_KIND_IN)    :: iac_ny          ! nx, ny of "2d" grid
 
+
```
It's best to not change whitespace if you can help it.

eclare108213 commented on components/mpas-seaice/src/shared/mpas_seaice_icepack.F:
```diff
@@ -367,6 +382,258 @@ subroutine init_column_itd(domain)
+    call MPAS_pool_get_config(domain % configs, "config_use_column_waves", config_use_column_waves)
+    call MPAS_pool_get_config(domain % configs, "config_use_floe_size_distribution",config_use_floe_size_distribution)
+   
+    ice_init_type='default' ! NOTE: this should probably be a namelist setting? how to define for restarts?
```
The namelist variable `ice_ic` in Icepack (and CICE) is overloaded in the sense that it can be used to specify whether a run will use a restart file or start from scratch, and if it's starting from scratch it specifies how the ice is configured, and if it's starting from a restart then it gives the name of the restart file (and yes I think this drives some of our users/developers a little batty).  MPAS-SI divides all of that functionality up into separate flags.  E.g. there's `config_initial_condition_type` in mpas_seaice_initialize.F, and there's `config_do_restart` and `runtype` in ice_comp_mct.F, so I think you'll just need to grab the ones you need and not worry about defining `ice_ic` separately.

eclare108213 commented on components/mpas-seaice/src/shared/mpas_seaice_icepack.F:
```diff
@@ -6548,6 +7007,13 @@ subroutine init_column_tracer_object_parent_indices(domain, tracerObject)
+    ! Floe size distribution
+    if (config_use_floe_size_distribution) then
+       do iFloeSize = 1, nFloeCategories
+          tracerObject % parentIndex(tracerObject % index_floeSizeDist + iFloeSize - 1) = 1 ! WHAT VALUE SHOULD THIS BE SET TO ????
```
I think this should be 0.  This designation indicates which quantity the tracer is being carried on or in.  In this case, the FSD is a tracer on ice area (so 0).  Snow characteristics in the code just above this are carried around in the snow volume (so 2).  The tracer design is pretty slick (makes advection easy) but it's a lot of bookkeeping.  How it works is described here, using melt ponds as an example: https://cice-consortium-icepack.readthedocs.io/en/main/science_guide/sg_tracers.html

erinethomas replied:
this is very helpful to understand the tracer stuff - thanks!

eclare108213 reviewed (commented):
Without looking too closely at the wave and coupling parts of this, I'd say you're on the right track.  Nothing here stands out to me as being seriously problematic, and I only have a few suggestions in the comments below.  This is looking great!  It's clearly a lot of work, already - thanks -

eclare108213 commented:
Also, it's not all that important, but you could change this to a draft PR (instead of "open"), to indicate that it's not intended to be merged at this point.  I ought to be able to make that change (?) but don't see how.

Commit:
a few minor fixes(such as typos and white spaces)
fix to floeSizeDist time dimension in initalization subroutine.

jonbob commented on components/mpas-seaice/bld/build-namelist:
```diff
@@ -416,6 +416,7 @@ my $RUN_STARTDATE          = "$xmlvars{'RUN_STARTDATE'}";
 my $START_TOD              = "$xmlvars{'START_TOD'}";
 my $RUN_REFDATE            = "$xmlvars{'RUN_REFDATE'}";
 my $CONTINUE_RUN           = "$xmlvars{'CONTINUE_RUN'}";
+my $wave_spec              = $xmlvars{'WAV_SPEC'};
```
you may need quotes around $xmlvars{'WAV_SPEC'}, as in the lines before this

erinethomas replied:
the quotes do not seem to matter -this part works fine as is - but I can add them for consistency.

jonbob commented on components/ww3/cime_config/config_compsets.xml:
```diff
@@ -24,14 +24,19 @@
+    <alias>DTEST-JRA1p5-WW3</alias>
+    <lname>2000_DATM%JRA-1p5_SLND_MPASSI_DOCN%SOM_DROF%JRA-1p5_SGLC_WW3%sp36x36_TEST</lname>
+  </compset>
+
```
the compset naming convention should be DTESTM instead of DTEST -- we use the "M" to differentiate mpassi usage from cice, which has a DTEST compset

jonbob commented on driver-mct/shr/seq_infodata_mod.F90:
```diff
@@ -782,6 +782,7 @@ SUBROUTINE seq_infodata_Init( infodata, nmlfile, ID, pioid, cpl_tag)
        infodata%iac_ny = 0
 
+
        infodata%nextsw_cday   = -1.0_SHR_KIND_R8
```
is there a reason for all the extra spaces in this file? I didn't see any real modified code

erinethomas replied:
no- this is a mistake left over from some old modifications that I have since removed.

jonbob commented:
@erinethomas -- looks like a lot of work! We'll also have to add some code to the coupler to pass the new field between ww3 and mpassi. I'm happy to help you with that part

Commit:
git rebase

Commit:
removing extra empy spaces

Commit:
infrastructure to wave - ice coupling. for example: adding fsd and wavespectra to seq_flds_mod.F90

Commit:
Define w2x_Sw_wavespec in ww3_cpl_indices.f90

Commit:
calcualte max floe diameter and add to i2x variables for wave coupling

Commit:
index_w2x_Sw_wavespec is now a function of frequency - should create individual
coupling variable for wave spectrum for every frequency

Commit:
fixing typos in ice_comp_mct for wave coupling variables and ensuring consistency with structure/variables within Registry.xml

Commit:
smll typo fix to wav_comp_mct

Commit:
fix loop to define Sw_wavespec for each freq value in wav_comp_mct.F90

Commit:
add EF to coupling and output namelist for WW3

Commit:
typo fix

Commit:
rename WaveSpectrum to 'WaveSpectra'

Commit:
add number of wave frequencies to drv_in namelist - cime requires the
number of wave frequencies used in order to define/initalize
the coupled wave spectrum variables (wave psctrum is a coupled 3D variable)

Commit:
remove empty spaces from prior commit

Commit:
add wav_nfreq and loop for waave spectra feilds to seq_flds_mod.F90

Commit:
full coupling of 3D wave spectra. (note code compiles but has not been tested at this point)

Commit:
FSD and Wave Ice coupling code complet + runs.
Note - accuracy of new code has not been completed yet.

Commit:
clean up structure of wave-realted mpas-si namelist settings.

Commit:
typo correction

Commit:
fix n-D aray for wave spectra in MPASSI

Commit:
bug fix to wave spectra definition in mpassi

Commit:
typo fix

Commit:
fix to wavespectra in ice comp_mct definition

Commit:
adding wave vars to daily and monthly output by dafault (mpas-seaice/cime_config/buildnml)

Commit:
fix to ice_ic definition in initialization of FSD.

Commit:
rebase icepack_step_therm2 call

Commit:
add nBioTracersLayer to steptherm2, comment out optional arguments for isotopes

Commit:
comment about CICE vs MPASSI use of Significant wave height

Commit:
minor syntax fix

Commit:
wave coupling definition

Commit:
define lateralHeatFlux Type

Commit:
define lateralIceMeltRate

Commit:
point to original icepack branch

Commit:
update a few wave vars

Commit:
bug fix to define wave spectra as zero in driver-mct namelist when WW3 is not used

Commit:
bug fix to wavefracture call- only if wave turned on.

Commit:
minor formatting updates

Commit:
removing duplicate variable definitions introduced in last rebase.

Commit:
bug fix to frequency calculation

Commit:
fix to buildnml so wave variables only used if WW3 is in compset.

Commit:
adding wave_spec_type to config options (set to 'random' if ww3 is coupled,
 and 'none' if ww3 is not used)

Commit:
add config_wave_spec_type to Registery.xml

Commit:
added initalization of icepack FSD/wave parameters to MPAS-SI values.

Commit:
update fsd related parameters in step therm 1 call

Commit:
add coupler infrastructure for remapping between MPAS-SI and WW3.

Commit:
call to ice_comp_init now includes wav_c2_ice argument

Commit:
add wav2ice to driver-mct/cime_config/config_component.xml

Commit:
add wav2ice smap to driver-mct/cime_config/namelist_definition_drv.xml

Commit:
define wave to ice remapping file for WQU225EC30to60

Commit:
typo fix for wav2ice smap

Commit:
fix to wav_c2_ice variable for wave to ice coupling

Commit:
remove print lines for testing

Commit:
add calculation for IceFloePerimeter per unit area

Commit:
move location of Hs - fixes issue with all zeros being passed from coupler to mpas-si

Commit:
add FloeSizePerimeter to defult output streams

Commit:
calculation for floe size dist perimeter in ice_export_mct ?

Commit:
fix to step therm2 for floediam (only uses this namelist constant is FSD is off)

Commit:
fix to fsd perimeter calcualtion. - note the location of this calculation will be moved to more approprate location in near future.

Commit:
fix to typo of 'config_floediam'

Commit:
add floesize analysis member.

eclare108213 commented:
@erinethomas 
Is there anything in this PR (or related changes to coupler, etc) that could change the answers when the FSD and waves are turned off?

proteanplanet commented:
And related to @eclare108213, is this at the point where we could run a back-to-back test with Icepack switched on, and the FSD switched on of off?

erinethomas commented:
@eclare108213 @proteanplanet - this PR is based on a code base that is now out of date. I will rebase today and run the ice tests.  That said, the tests I am running now (D case w/ and w/o FSD) are working. Again, these fsd tests are using an old code base - so I will be more comfortable answering this after I rebase the code (top of priority list will be done by EOB today.

erinethomas commented:
Follow up comment: If FSD is tuned off and WW3 IS NOT USED. This PR and the FSD code **shouldn't** change anything (but the ice testing will confirm this)

erinethomas committed:
Merge branch 'WaveIceCoupling_icepackintegration' into dev_WaveIceCoupling

erinethomas commented:
closed this PR since the wave ice coupling will go right into v3 master
